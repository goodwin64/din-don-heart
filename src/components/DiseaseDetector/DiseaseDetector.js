import { Component } from 'react';
import { connect } from 'react-redux';

import {
  ecgLettersPT,
  onDiseaseResultPT,
} from '../../helpers/proptypes.helper';
import getDisease from '../../helpers/getDisease';
import { onDiseaseResultServerAnalysis } from '../../actions/onDiseaseResult';

const wait = seconds => new Promise(resolve => setTimeout(resolve, seconds));

class DiseaseDetector extends Component {
  static propTypes = {
    ecgLettersDetailed: ecgLettersPT.isRequired,
    onDiseaseResult: onDiseaseResultPT.isRequired,
  };

  componentDidMount() {
    this.sendEcgForAnalysis(this.props.ecgLettersDetailed);
  }

  componentWillReceiveProps(props) {
    if (props.ecgLettersDetailed !== this.props.ecgLettersDetailed) {
      this.sendEcgForAnalysis(props.ecgLettersDetailed);
    }
  }

  sendEcgForAnalysis = (ecgLettersDetailed) => {
    const fakeServerResponse = wait(0).then(() => ({ json: () => getDisease(ecgLettersDetailed) }));
    const realServerResponse = fetch(`https://randomuser.me/api/?results=10&ecgLetters=${ecgLettersDetailed}`);

    Promise.race([
      fakeServerResponse,
      realServerResponse,
    ]).then(result => result.json())
      .then((result) => {
        const resultStringified = JSON.stringify(result);
        this.props.onDiseaseResult(resultStringified);
      })
      .catch((err) => {
        console.error('Error in Disease Detector', err);
      });
  };

  render() {
    return null;
  }
}

export default connect(null, {
  onDiseaseResult: onDiseaseResultServerAnalysis,
})(DiseaseDetector);
