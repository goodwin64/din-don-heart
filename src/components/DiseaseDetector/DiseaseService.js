import getDisease from '../../helpers/getDisease';

const wait = seconds => new Promise(resolve => setTimeout(resolve, seconds));

export default class DiseaseService {
  constructor(onDiseaseResult) {
    this.onDiseaseResult = onDiseaseResult;
  }

  sendEcgForAnalysis = (ecgLettersDetailed) => {
    const fakeServerResponse = wait(0).then(() => ({ json: () => getDisease(ecgLettersDetailed) }));
    const realServerResponse = fetch(`https://randomuser.me/api/?results=10&ecgLetters=${ecgLettersDetailed}`);

    Promise.race([
      fakeServerResponse,
      realServerResponse,
    ]).then(result => result.json())
      .then((result) => {
        this.onDiseaseResult(JSON.stringify(result));
      })
      .catch((err) => {
        console.error('Error in DiseaseService', err);
      });
  };
}
