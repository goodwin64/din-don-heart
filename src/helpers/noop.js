export function noop() {
  // named export used for test purposes only
}

export default { noop };
