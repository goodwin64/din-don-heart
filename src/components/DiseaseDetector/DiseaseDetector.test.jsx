import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import noop from '../../helpers/noop';
import { DiseaseDetector } from './DiseaseDetector';

// TODO: add async action and test it
describe('DiseaseDetector component', () => {
  // let mockDiseaseService;
  // let spySendEcgForAnalysis;
  // let spyOnDiseaseResult;

  beforeEach(() => {
    // spySendEcgForAnalysis = jest.fn();
    // spyOnDiseaseResult = jest.fn();
    // mockDiseaseService = class {
    //   constructor() {
    //     this.onDiseaseResult = spyOnDiseaseResult;
    //   }
    //
    //   sendEcgForAnalysis(params) {
    //     spySendEcgForAnalysis(params);
    //     this.onDiseaseResult();
    //   }
    // };
  });

  it('should render nothing', () => {
    const wrapper = shallow(<DiseaseDetector
      ecgLettersDetailed="A"
      onDiseaseResult={noop}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // it('should run onDisease stuff after mount', () => {
  //   expect(spySendEcgForAnalysis).not.toBeCalled();
  //   expect(spyOnDiseaseResult).not.toBeCalled();
  //
  //   shallow(<DiseaseDetector
  //     ecgLettersDetailed="A"
  //     onDiseaseResult={noop}
  //     DiseaseService={mockDiseaseService}
  //   />);
  //
  //   expect(spySendEcgForAnalysis).toBeCalledWith('A');
  //   expect(spyOnDiseaseResult).toBeCalled();
  // });
  //
  // it('should not run onDisease stuff if ECG letters are the same', () => {
  //   expect(spySendEcgForAnalysis).not.toBeCalled();
  //   expect(spyOnDiseaseResult).not.toBeCalled();
  //
  //   const wrapper = shallow(<DiseaseDetector
  //     ecgLettersDetailed="A"
  //     onDiseaseResult={noop}
  //     DiseaseService={mockDiseaseService}
  //   />);
  //
  //   expect(spySendEcgForAnalysis).toHaveBeenCalledTimes(1);
  //   expect(spyOnDiseaseResult).toHaveBeenCalledTimes(1);
  //
  //   wrapper.setProps({
  //     ecgLettersDetailed: 'A',
  //   });
  //
  //   expect(spySendEcgForAnalysis).toHaveBeenCalledTimes(1);
  //   expect(spyOnDiseaseResult).toHaveBeenCalledTimes(1);
  //
  //   wrapper.setProps({
  //     ecgLettersDetailed: 'B',
  //   });
  //
  //   expect(spySendEcgForAnalysis).toBeCalledWith('B');
  //   expect(spySendEcgForAnalysis).toHaveBeenCalledTimes(2);
  //   expect(spyOnDiseaseResult).toHaveBeenCalledTimes(2);
  // });
});
