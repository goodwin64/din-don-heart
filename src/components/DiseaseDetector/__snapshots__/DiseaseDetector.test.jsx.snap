// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiseaseDetector component should render nothing 1`] = `""`;
